osf/models/collection_submission.py

```diff
@@ -0,0 +1,255 @@
+from osf.models.mixins import TaxonomizableMixin
+from osf.utils.permissions import ADMIN
+from website.util import api_v2_url
+from website.search.exceptions import SearchUnavailableError
+from osf.utils.workflows import CollectionSubmissionsTriggers, CollectionSubmissionStates
+from website import mails, settings
+from osf.utils.machines import CollectionSubmissionMachine
+from django.db.models.signals import post_save
+from django.dispatch import receiver
+
+
+logger = logging.getLogger(__name__)
+
+
+class CollectionSubmission(TaxonomizableMixin, BaseModel):
+    primary_identifier_name = 'guid___id'
+
+    class Meta:
+        order_with_respect_to = 'collection'
+        unique_together = ('collection', 'guid')
+
+    collection = models.ForeignKey('Collection', on_delete=models.CASCADE)
+    guid = models.ForeignKey('Guid', on_delete=models.CASCADE)
+    creator = models.ForeignKey('OSFUser', on_delete=models.CASCADE)
+    collected_type = models.CharField(blank=True, max_length=127)
+    status = models.CharField(blank=True, max_length=127)
+    volume = models.CharField(blank=True, max_length=127)
+    issue = models.CharField(blank=True, max_length=127)
+    program_area = models.CharField(blank=True, max_length=127)
+    school_type = models.CharField(blank=True, max_length=127)
+    study_design = models.CharField(blank=True, max_length=127)
+    machine_state = models.IntegerField(
+        choices=CollectionSubmissionStates.int_field_choices(),
+        default=CollectionSubmissionStates.IN_PROGRESS,
+    )
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.state_machine = CollectionSubmissionMachine(
+            model=self,
+            active_state=self.state,
+            state_property_name='state'
+        )
+
+    @property
+    def state(self):
+        return CollectionSubmissionStates(self.machine_state)
+
+    @property
+    def is_moderated(self):
+        return self.collection.provider and self.collection.provider.reviews_workflow == 'post-moderation'
+
+    @state.setter
+    def state(self, new_state):
+        self.machine_state = new_state.value
+
+    def _notify_moderated_pending(self, event_data):
+        for contributor in self.guid.referent.contributors:
+            mails.send_mail(
+                to_addr=contributor.username,
+                mail=mails.COLLECTION_SUBMISSION_SUBMITTED,
+                user=contributor,
+                submitter=self.creator,
+                is_initator=self.creator == contributor,
+                is_admin=self.guid.referent.has_permission(contributor, ADMIN),
+                is_registered_contrib=contributor.is_registered,
+                collection=self.collection,
+                claim_url=contributor.unclaimed_records.get(self.guid.referent._id, None),
+                node=self.guid.referent,
+                domain=settings.DOMAIN,
+                osf_contact_email=settings.OSF_CONTACT_EMAIL,
+            )
```

Comment:
How much of this context is shared across emails? Could this be refactored into some sort of `generate_email_context` helper function?

Reply:
let's put this off until product can confirm there are no changes.